### ISSUE TITLE
nxos_snmp_server is not gathering enabled BGP and OSPF traps

### ISSUE THREAD

**Issue by ledgley** [ansible-collections/cisco.nxos#465]
##### SUMMARY
nxos_snmp_server is not gathering enabled BGP and OSPF traps. These traps can be enabled on their own without any sub commands.

##### ISSUE TYPE
- Bug Report

##### COMPONENT NAME
cisco.nxos.nxos_snmp_server

##### ANSIBLE VERSION
```
ansible 2.10.17
  config file = /Users/edgleyl/.ansible.cfg
  configured module search path = ['/Users/edgleyl/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /Users/edgleyl/venv/brew-ten/lib/python3.9/site-packages/ansible
  executable location = /Users/edgleyl/venv/brew-ten/bin/ansible
  python version = 3.9.12 (main, Mar 26 2022, 15:51:15) [Clang 13.1.6 (clang-1316.0.21.2)]
```

##### COLLECTION VERSION
```
(brew-ten) edgleyl@edgleyl-JSS1242 compliance % ansible-galaxy collection list | grep nxos
cisco.nxos             2.8.1
```

##### CONFIGURATION
```
(brew-ten) edgleyl@edgleyl-JSS1242 compliance % ansible-config dump --only-changed
DEFAULT_CALLBACK_WHITELIST(/Users/edgleyl/.ansible.cfg) = ['timer']
DEFAULT_STDOUT_CALLBACK(/Users/edgleyl/.ansible.cfg) = yaml
HOST_KEY_CHECKING(/Users/edgleyl/.ansible.cfg) = False
INTERPRETER_PYTHON(/Users/edgleyl/.ansible.cfg) = /Users/edgleyl/venv/brew-ten/bin/python
PERSISTENT_COMMAND_TIMEOUT(/Users/edgleyl/.ansible.cfg) = 180
```

##### OS / ENVIRONMENT
NXOS 7.0(3)I7(3)

##### STEPS TO REPRODUCE
This is the device configuration snippet
```
snmp-server enable traps bgp
snmp-server enable traps ospf
snmp-server enable traps config ccmCLIRunningConfigChanged
snmp-server enable traps snmp authentication
```
This is the task
```yaml
- name: GATHER DEVICE SNMP CONFIG
  cisco.nxos.nxos_snmp_server:
    state: gathered
  register: output
```

##### EXPECTED RESULTS
Expect OSPF and BGP to be gathered, as they are enabled
Eg.
```yaml
  output:
    changed: false
    failed: false
    gathered:
      traps:
        bgp: true
        config:
          ccmCLIRunningConfigChanged: true
        ospf: true
        snmp:
          authentication: true
```

##### ACTUAL RESULTS
This is the resulting gathered data
```yaml
  output:
    changed: false
    failed: false
    gathered:
      traps:
        config:
          ccmCLIRunningConfigChanged: true
        snmp:
          authentication: true
```

**Comment by ledgley** [comment]
NXOS isn't showing all configured SNMP traps using `show running-config`, however using `show snmp trap` does.

Eg.
```
snmp-server enable traps bgp
snmp-server enable traps ospf
snmp-server enable traps config ccmCLIRunningConfigChanged
snmp-server enable traps snmp authentication
```

```
ausp-fab-ls003# show snmp trap | i Yes
entity               : entity_mib_change               Yes
entity               : entity_module_status_change     Yes
entity               : entity_power_status_change      Yes
entity               : entity_module_inserted          Yes
entity               : entity_module_removed           Yes
entity               : entity_unrecognised_module      Yes
entity               : entity_fan_status_change        Yes
entity               : entity_power_out_change         Yes
link                 : linkDown                        Yes
link                 : linkUp                          Yes
link                 : extended-linkDown               Yes
link                 : extended-linkUp                 Yes
link                 : cieLinkDown                     Yes
link                 : cieLinkUp                       Yes
link                 : delayed-link-state-change       Yes
rf                   : redundancy_framework            Yes
license              : notify-license-expiry           Yes
license              : notify-no-license-for-feature   Yes
license              : notify-licensefile-missing      Yes
license              : notify-license-expiry-warning   Yes
upgrade              : UpgradeOpNotifyOnCompletion     Yes
upgrade              : UpgradeJobStatusNotify          Yes
rmon                 : risingAlarm                     Yes
rmon                 : fallingAlarm                    Yes
rmon                 : hcRisingAlarm                   Yes
rmon                 : hcFallingAlarm                  Yes
config               : ccmCLIRunningConfigChanged      Yes
snmp                 : authentication                  Yes
entity               : entity_sensor                   Yes
entity               : cefcMIBEnableStatusNotification Yes
generic              : coldStart                       Yes
generic              : warmStart                       Yes
storm-control        : cpscEventRev1                   Yes
link                 : cErrDisableInterfaceEventRev1   Yes
pim                  : pimNeighborLoss                 Yes
ospf-OVERLAY         : OSPF base traps                 Yes
ospf-OVERLAY         : OSPF LSA                        Yes
ospf-UNDERLAY        : OSPF base traps                 Yes
ospf-UNDERLAY        : OSPF LSA                        Yes
BGP-64602            :                                 Yes
```

Then I disable the linkDown trap with `no snmp-server enable  trap link linkDown` and see 

```
entity               : entity_mib_change               Yes
entity               : entity_module_status_change     Yes
entity               : entity_power_status_change      Yes
entity               : entity_module_inserted          Yes
entity               : entity_module_removed           Yes
entity               : entity_unrecognised_module      Yes
entity               : entity_fan_status_change        Yes
entity               : entity_power_out_change         Yes
link                 : linkUp                          Yes
link                 : extended-linkDown               Yes
link                 : extended-linkUp                 Yes
link                 : cieLinkDown                     Yes
link                 : cieLinkUp                       Yes
link                 : delayed-link-state-change       Yes
rf                   : redundancy_framework            Yes
license              : notify-license-expiry           Yes
license              : notify-no-license-for-feature   Yes
license              : notify-licensefile-missing      Yes
license              : notify-license-expiry-warning   Yes
upgrade              : UpgradeOpNotifyOnCompletion     Yes
upgrade              : UpgradeJobStatusNotify          Yes
rmon                 : risingAlarm                     Yes
rmon                 : fallingAlarm                    Yes
rmon                 : hcRisingAlarm                   Yes
rmon                 : hcFallingAlarm                  Yes
config               : ccmCLIRunningConfigChanged      Yes
snmp                 : authentication                  Yes
entity               : entity_sensor                   Yes
entity               : cefcMIBEnableStatusNotification Yes
generic              : coldStart                       Yes
generic              : warmStart                       Yes
storm-control        : cpscEventRev1                   Yes
link                 : cErrDisableInterfaceEventRev1   Yes
pim                  : pimNeighborLoss                 Yes
ospf-OVERLAY         : OSPF base traps                 Yes
ospf-OVERLAY         : OSPF LSA                        Yes
ospf-UNDERLAY        : OSPF base traps                 Yes
ospf-UNDERLAY        : OSPF LSA                        Yes
BGP-64602            :                                 Yes
```

Tested this on `7.0(3)I7(3)` and `9.3(8)`

**Comment by NilashishC** [comment]
@ledgley could you please share which traps do not appear in running-config in an enabled state?

**Comment by ledgley** [comment]
@NilashishC this gets a bit more complicated than the original comment, it may warrant opening a second issue. I updated the comment above to include the commands used to show the output.

In the comment above I have shown what is present in running-config with `show running-config | i 'enable traps'`

Then, I show which traps are enabled with `show snmp trap | i Yes`

Of all the enabled traps in the second output, only the four in the first output are in running-config.

That said, those that are not shown in running-config but are enabled, are enabled by default. Default config is not shown in running-config. For example, if I use `no snmp-server enable traps upgrade UpgradeJobStatusNotify` this trap is disabled, but the 'no' line shows in running-config.
```
ausp-fab-ls003# show run | i UpgradeJobStatusNotify
no snmp-server enable traps upgrade UpgradeJobStatusNotify

ausp-fab-ls003# show snmp trap | i UpgradeJobStatusNotify
upgrade              : UpgradeJobStatusNotify          No
```

In testing this, I noticed that if the default config is disabled (ie. UpgradeJobStatusNotify is disabled) this is not gathered by the module. Nor can I change the status of the default config using the module eg.

```yaml
    - name: MERGE DEVICE SNMP CONFIG
      cisco.nxos.nxos_snmp_server:
        state: merged
        config:
          traps:
            upgrade:
              UpgradeJobStatusNotify: False
```

```
TASK [MERGE DEVICE SNMP CONFIG] **********************************************************************************************************************************************
fatal: [ausp-fab-ls003]: FAILED! => changed=false
  msg: 'Unsupported parameters for (cisco.nxos.nxos_snmp_server) module: UpgradeJobStatusNotify found in config -> traps -> upgrade. Supported parameters include: enable, upgradeJobStatusNotify, upgradeOpNotifyOnCompletion'
```

It makes sense to not gather default configuration however we still need to be able to turn all of the snmp trap knobs.

**Comment by NilashishC** [comment]
@ledgley Ah, I get it now. The `upgradejobstatusnotify` traps is in fact supported. The task you shared uses an incorrect key name (should be lowercase `u`). Please refer to the doc for all available options. https://docs.ansible.com/ansible/latest/collections/cisco/nxos/nxos_snmp_server_module.html#parameter-config/traps/upgrade/upgradeJobStatusNotify

However, as of today, you'd be able to configure it, but not gather it as facts. This is because the show command for this module has a "starts with" snmp-server filter, which means that negated forms for the commands are not read from the device.
This would be a considerably large change, since I'd have go through all traps and check which one appears in it's "no" form in running-config and which one doesn't. Could you please open a separate ticket for this? I'll make sure that this get addressed. Thank you for bringing this to our notice. Really appreciate it!

As for the original issue you mentioned in this bug report, it should be fixed with the patch in #466.

**Comment by ledgley** [comment]
Thanks @NilashishC !
